### ISSUE TITLE
Design a formatter extension API

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#31691]
Currently the formatter is only configurable by a predefined set of options. We should design a public API allowing extensions to the formatter, so automatic formatting operations triggered by user actions in the IDE are better able to handle uncommon coding patterns. Projects using code formatting styles not supported by the current set of options could install a formatter extension to provide the missing functionality.

* The extension API should support `FormattingAnalyzer` in the Code Style layer, which runs during compilation without dependencies on the Workspaces layer.
* The extension API should support the following formatting scenarios:
    * Formatting in columns (#28729)
    * Formatting of preprocessor directives (#12306)

**Comment by CyrusNajmabadi** [comment]
I'm interested in this (especially around the idea of easily being able to add formatting analyzers to ensure things like Roslyn/IDE consistency).  When you do this design work, can you include me? Thanks!

**Comment by sharwell** [comment]
The first part of the evaluation will be determining if it will be sufficient to allow extensions to add items to this list:

https://github.com/dotnet/roslyn/blob/def06e4bd7f53ab7989bd0c86505aac6f80d9e95/src/Workspaces/CSharp/Portable/Formatting/CSharpSyntaxFormattingService.cs#L33-L46

**Comment by CyrusNajmabadi** [comment]
@sharwell Makes sense to me.  I'm of two minds about it.  

1. it's a good idea since this is the same system the rest of hte formatter is built off of.
2. it's not clear to me if the current approach to a formatting rule is one we want users to use.  From my own experience trying to write a formatting rule, it's both exceptionally complex, and often very hard to diagnose what went wrong.  

There's a part of me that just likes the idea of allowing someone to add just simple tree-transforms.  We could even make it more narrow and just allow trivial updates.  Or allow different types of "rules" for these more common cases.

**Comment by lostmsu** [comment]
@sharwell currently, `IndentBlockFormattingRule` handles all types of blocks internally. So to cover #24129 , it has to be refactored. Just adding a new rule to that list will cause conflicting processing.

**Comment by JC3** [comment]
This sounds neat but... effectively it seems the same as adding better options to the editor settings then writing a .editorconfig file, except this is a lot more effort for the user who just wants to e.g. not indent namespaces.

The C++ editor has a mature and very satisfactory set of options, and could serve as a very good basis of options to provide. Then a .editorconfig file would basically be the same as a "formatter extension", except written in a different language (it'd express the same intents, as a settings document instead of a C# source file).

**Comment by CyrusNajmabadi** [comment]
@JC3 It's unclear to me what you're asking for here.

**Comment by sharwell** [comment]
> effectively it seems the same as adding better options to the editor settings then writing a .editorconfig file, except this is a lot more effort for the user who just wants to e.g. not indent namespaces.

No, there is a huge difference in quality expectations between the two approaches. In is algorithmically impossible for us to test all the combinations of options we have today, and every single option we add _at minimum_ doubles the target test space. On the other hand, an extensibility layer means individual projects can design a single set of options with known values, bringing the test space down to one known and well-tested set. Also, bugs can be addressed by that team if/when they arise, and those fixes can be made quickly with understanding of the specific needs of a single project (or set of projects). Changing built-in options is nearly impossible, even in cases where most users would disagree with them.

**Comment by rjgotten** [comment]
> The formatter extension API seems cool but realistically how much usage is it going to get? It seems like it will primarily be users adding feature that are similar to the C++ editor and then, once somebody nails them all in one extension and makes that publicly available, that's about it for that -- there will be one or two "Better C# Formatting" plugins in the extensions marketplace and nobody will really need to touch the formatter extension API again (I hypothesize). It's a lot of extra layers between the current state and "add spaces around binary operators" or whatever.

FWIW even if there were to be only one or two big-name formatting plugins out there, I would (almost) kill for the C# equivalent to ESLint. The current formatter just smacks of limitations. Both in what rules and rule configurations it supports _and_ in how much actual development effort goes into it.

Not having control over how the editor formats your code is already one of the top-ranking annoyances you can saddle developers with. Visual Studio in a way makes it _worse_ by also not even giving you full control over _if_ or _when_ it formats your code. E.g. #51389 -- why are we only able to disable auto-mangling of aligned white-space for assignments within declaration assignments, but specifically _not_ for the much, _much_ more useful cases of object initializer syntax or pattern matching?

In any environment that _wouldn't_ have monopolistic lock-in this kind of thing would build in annoyance to the point where it would viably become a reason to drop the offering and seek alternatives instead. But you don't get that with Visual Studio or VSCode. Because its Roslyn all the way down and the formatting API is clamped down.

At the very least, a succesful third party running dedicated maintenance time on a formatting plugin, means you can _actually get shit done_ and offer _actual solutions_ -- even to small nitpicks. Because small as they may _seem_ to the Roslyn maintainers on their grand scale, they _will_ build major frustration with developers that have no recourse but to deal with the fallout, day after day; week after week. Death of a thousand paper cuts.

One need only look at ESLint to see that the principle behind a dedicated team working on a formatter _actually works_. And in the small cases where it _doesn't_ a small - maybe somewhat hacked - formatter can be used as a stop-gap. Better yet; separate pluggable formatters can (and in many cases _are_ !!) used as proof-of-concepts before being formally adopted into the core ESLint rule sets.

**Comment by jez9999** [comment]
Would like to see this get implemented.  I'd like `#pragma`s to be indented along with my code (yeah, I know they're scopeless, but not indenting them still looks horrible).

**Comment by tycobbb** [comment]
please.

**Comment by jez9999** [comment]
IMHO this should be bumped up to high priority now.  The formatting functionality is rather limited to a set of built-in options, everything bundled under the "IDE0055 Fix formatting" rule so you don't always know what exactly is wrong with the formatting, and the formatting functionality is failing to keep up with new C# language features properly.  Two examples I encountered upon upgrading my project to .NET 8:

https://github.com/dotnet/roslyn/issues/71218
https://github.com/dotnet/roslyn/issues/71219

I mean, not even having a way to get the formatter to do `[..list]` instead of `[.. list]` with an awkward space in?  Please, guys, overhaul the formatting engine.  Make it much more flexible, don't group everything under one IDE0055 rule, and make it easy for plugins to customize!

**Comment by CyrusNajmabadi** [comment]
> IMHO this should be bumped up to high priority now

This is not a question of priority. The main issue is that it is not apparent what the design would be for a formatting engine that was fully flexible like it want, while also being simple to extend, while also being fast enough. 

> and the formatting functionality is failing to keep up with new C# language features properly.

Formatting is always updated to handle new constructs. 

> to do [..list]

This is not a formatting style the team considers acceptable for collection expressions. Like with the corresponding pattern (`x is [var start, .. var middle, var end]`), it is very expected that the space be present here. 

> Make it much more flexible, ... and make it easy for plugins to customize!

The question is how to do this while also being fast enough, and having API be understandable and flexible enough. 

As a starting point, how do you design such that *any* behavior be customizable?

**Comment by jez9999** [comment]
> This is not a formatting style the team considers acceptable for collection expressions. Like with the corresponding pattern (`x is [var start, .. var middle, var end]`), it is very expected that the space be present here.

Well sorry, but why should "the team" dictate what is "acceptable"?  In JS, the standard for the spread operator doesn't use a space, I prefer it without a space, and it doesn't have a space on the "What's new in C# 12 page" examples:

https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#collection-expressions

**Comment by CyrusNajmabadi** [comment]
> Well sorry, but why should "the team" dictate what is "acceptable"?

That's our job.  We design the language and implement the features.

>  In JS, 

We're not JS.  That's a separate design group and ecosystem. 

> and it doesn't have a space on the "What's new in C# 12 page" examples:

Thanks.  Ill fix.

**Comment by jez9999** [comment]
> > Well sorry, but why should "the team" dictate what is "acceptable"?
> 
> That's our job. We design the language and implement the features.

That's not the same as dictating people's exact syntax though, is it?  That's why there are a ton of options as to what kind of formatting will be used when it comes to spacing.

And yeah, you'll "fic" the page.  Whatever.  The fact that it didn't have a space there in the first place shows that I'm not the only one thinging that style looks best.  Why would you want to take away the option to customize this?  Please consider adding such an option.

**Comment by CyrusNajmabadi** [comment]
> That's not the same as dictating people's exact syntax though, is it? 

In a real sense yes.  If you want to use these tools (which are optional).  Since C# 1.0 we've had many features with no control over their formatting.  That's been the *normal* case for C# for 20+ years now.  It's the *rare* feature that gets an extension point.  And generally only when there is enough feedback for it.

> And yeah, you'll "fic" the page

Thanks. The change was taken.  It'll be updated soon.  Thanks for bringing this to our attention.

> Why would you want to take away the option to customize this? 

Generally, because customization adds complexity and maintenance costs.  It also bifurcates the ecosystem.   So, generally, the position is that if you want to use these (optional) tools, you're opt'ing into the formatting choices they make (like many linters).  We totally get that that may not be desired, which is why it can be trivially disabled so you can format your code however you want.

**Comment by CyrusNajmabadi** [comment]
> Please consider adding such an option.

We will definitely consider adding such an option.  Thanks.

**Comment by jez9999** [comment]
> > That's not the same as dictating people's exact syntax though, is it?
> 
> In a real sense yes. If you want to use these tools (which are optional). Since C# 1.0 we've had many features with no control over their formatting. That's been the _normal_ case for C# for 20+ years now. It's the _rare_ feature that gets an extension point. And generally only when there is enough feedback for it.

So why did you even allow the spread operator not to have a space after it?  Why isn't `[..list]` invalid syntax?  And who wrote that style in the "What's new in C# 12" page, apparently thinking it looked fine?

> Generally, because customization adds complexity and maintenance costs. It also bifurcates the ecosystem. So, generally, the position is that if you want to use these (optional) tools, you're opt'ing into the formatting choices they make (like many linters).

This is ridiculous.  Whether or not a space is there doesn't "bifurcate the ecosystem", people can obviously understand the code  just fine whether the space is there or not.

> We totally get that that may not be desired, which is why it can be trivially disabled so you can format your code however you want.

Only by putting a big ugly couple of `#pragma` lines around various individual lines of code, which is highly undesirable.

What do you think about this @sharwell ?  Is it reasonable that C# formatting options already provides [22 different customizations](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/csharp-formatting-options#spacing-options) for customizing where spaces should be placed because people have different C# styles, but that the formatter always forces a space after a spread operator with no customization?

**Comment by CyrusNajmabadi** [comment]
> So why did you even allow the spread operator not to have a space after it?

Because the language doesn't care.  There are almost no places where whitespace is relevant to hte langauge.  However, the reverse is true of the formatter/linter.  Where whitespace is relevant almost everywhere :)

> Whether or not a space is there doesn't "bifurcate the ecosystem", 

It definitely does.  And it's why some ecosystems (like Go) just have removed options entirely from formatting with tools like `gofmt`.  Precisely because this causes so many religious wars.  Heck, just see the passion being demonstrated here :)

> Is it reasonable that C# formatting options already provides [22 different customizations](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/csharp-formatting-options#spacing-options)

Note: by my rough count, there are around 300 non customized places (or even more depending on how you want to count certain variations).  Non-customization is the *norm* for the language.

**Comment by jez9999** [comment]
Ability to have my own custom code style is actually one of the reasons I like C#.  Attempts to move it towards One True Style are extremely frustrating and I am 100% opposed to that kind of prescriptivism.

Frankly, what this discussion demonstrates to me is that the code formatting should be a lot more easily customizable with plugins so that this kind of thing becomes a non-issue.  The enforcement for spaces after the spread operator should be implemented with a plugin that I could just remove, or modify to have the behaviour I want.  Why isn't that the end goal?

**Comment by CyrusNajmabadi** [comment]
> Ability to have my own custom code style is actually one of the reasons I like C#.

You can 100% have your own custom code style.  The language and tools allow for it.  But the formatter itself is not extensible in this fashion (and we don't know of a design currently that would allow it to be).  It's something we're interested in, but haven't found worthwhile to do yet.  As such, for the last 20 years, that has meant the majority of the language just formats in a single way by default.  

> is that the code formatting should be a lot more easily customizable with plugins so that this kind of thing becomes a non-issue. 

If you have a design that allows for that to happen, and can be fast and maintainable, def let ys know.

> Why isn't that the end goal?

It is a goal.  We just are unaware of a system that can effectively provide that in a cohesive fashion with enough performance.  For example, how do you expression constraints between the different formatting plugins?  How does one override another?  How does one suppress another?  How do you make sense of it when tehre are hundreds (or thousands of them in play).  Note that thsi is the state of hte formatter today (which is internal).  And even in this state it is near impossible to touch any part of it without breaking some formatting in some place in an unintended fashion.  That's a deep challenge here.

**Comment by jez9999** [comment]
Out of interest, why does your team think whitespace after the spread operator is a good idea?  It's a unary operator, and every other symbolic unary operator in C# that I can think of is used without whitespace.  This: `if (! true) { ... }` looks silly.  Which is probably why I can find various examples scattered around MS Docs of the spread operator being used without whitespace.  IMHO you guys should rethink whether your preferred coding style is right.

**Comment by CyrusNajmabadi** [comment]
> It's a unary operator,

It's not a unary operator.  It's not an operator at all.

> why does your team think whitespace after the spread operator is a good idea?

Spacing strongly indicates it's not binding to, or affecting the thing right next to it (like an operator, or unary operator would).

For example, if this was a unary operator, then:

```
.. x ? y z
```

should be thought of as `(.. x) ? y : z`

But that's *100%* not the semantics or intuition for what this is.  As above, it's not an operator.  And that code example should be interpreted and thought of as if it was `.. (x ? y : z)`.  As such, our strong syntactic preference is not to have these abut as it's honestly very misleading.

Note that `!` is, as you mentioned, an operator.  And using the logic above *should* be next to the next hting.  That because `!x ? y : z` should be thought of as `(!x) ? y : z` and *not* `! (x ? y : z)`.

**Comment by jez9999** [comment]
Not an operator?  Yikes.  There are places all over MS Docs and tutorials websites calling it a "spread operator".

**Comment by CyrusNajmabadi** [comment]
> Not an operator? Yikes

Definitely not 'yikes'.  As per above, it would be very bad if it was.  You'd have to write things out as `.. (x ? y : z) ` instead of the desired `.. x ? y : z` (and the same with any other sort of expression case.  It's not a unary operator.  It never was, and we have no desire for it ever to be.

**Comment by CyrusNajmabadi** [comment]
> just one with low precedence. 

Even if that was how we viewed it (and we we don't), pperators with low precedence should not be formatted as if they tightly bind to an expression and thus have higher precedence than the code around them.  That's exactly why spacing here is important.  So even with that view, it woudl justify this spacing.

**Comment by jez9999** [comment]
If it's not an operator, what is it?

**Comment by CyrusNajmabadi** [comment]
> It operates on an enumerable operand, no?

It doesn't operate at all :)  It's not an expression.  It's just part of the collection syntax.  It's like saying that `,` is an operator in a collection-expression that adds the next thing.

**Comment by jez9999** [comment]
In that case there's a bunch of documentation out there that's very misleading.

**Comment by CyrusNajmabadi** [comment]
> If it's not an operator, what is it?

It's a type of collection expression element.  We currently have two types of elements, and we're planning on adding at least one more (but possibly two more).

We have in C# 12:

1. Expression elements.  Like `[a, b, c]`.  That is just an expression.
2. Spread elements.  Like `[.. a, .. b ? c : d]`.  It's a `..` followed by any expression.  There is no precedence or operating.  

In c# 13 we intend to add:

1. Dictionary elements.  Like `[a: b]`.  This is an expression followed by a colon, followed by an expression.  As with `..` there is no 'operating' on anything.  

We are also potentially considering a "conditional element".  Which would be something like `[? expr]` where `expr` would only be added if it was non-null.  Similar to `..` this is not operating on anything, and it would be expected that if you had a complex express on the right of this, that `?` was distinct and not bound tightly like a unary operator.

**Comment by CyrusNajmabadi** [comment]
> In that case there's a bunch of documentation out there that's very misleading.

Sounds like it.  We'll fix it up.  Thanks!

**Comment by CyrusNajmabadi** [comment]
> From a programmer's point of view, that feels like a distinction without a difference

I can't help with that.  Some things are operators, others are not.  And we treat them accordingly.  If it was an operator, you'd be able to use it anywhere were operators are allowed.  That is def not ok or appropriate (especially as it would *actually* conflict with the real operator we have, the range `..` operator).

If someone comes and tells me they think that commas are operators in our language, i'll tell them the same. 

> There's obviously precedence because you need to decide whether it's

No.  Precedence literally doesn't exist here.  Precedence occurs with constructs where the interpretation is affected by what comes afterwards.  for example, if i have `x + a * b` then the precedence of `+` matters vs `*` because it will impact how `a` is treated (should it be pulled into the `+` or not).   That literally doesn't occur here. In no way, shape, or form does the `..` impact parsing of anything.  

> Yeah... this is why some people say you can have too much syntactic sugar. 

Honestly, that's on them.  Dictionaries are a core collection type (they make up nearly 20 of all collections made the last time we surveyed the ecosystem (including nuget).  Being able to create them has always been part of hte collection-expression design.  We just pulled it out of 12 for time and resourcing needs.

> Maybe it's time to stop and let the language be more verbose?

As I said, `?` is something we're *considering*, not something we're taking.  If you want to give feedback on the design of the language, please give that in the design discussions happening on those topics at dotnet/csharplang.

**Comment by CyrusNajmabadi** [comment]
> the spread operator fits very naturally right next to its "operands" (for want of a better word).

Then you're welcome to format your code that way.  The language allows for it.  :)

**Comment by jez9999** [comment]
> > the spread operator fits very naturally right next to its "operands" (for want of a better word).
> 
> Then you're welcome to format your code that way. The language allows for it. :)

Yeah, but I shouldn't have to waste 2 lines wrapping it with `#pragma` just to stop the formatter complaining about it (thereby losing all other formatter stuff on that line as well), or even worse, disable all formatting analysis.  If you're honestly suggesting I do that everywhere I use the spread operator, I've gotta assume you're being sarcastic.  If the language designers saw fit to allow for it, the formatter should at least be able to be configured not to enforce spacing rules of any kind on the spread operator.

> No. Precedence literally doesn't exist here. Precedence occurs with constructs where the interpretation is affected by what comes afterwards. for example, if i have `x + a * b` then the precedence of `+` matters vs `*` because it will impact how `a` is treated (should it be pulled into the `+` or not). That literally doesn't occur here. In no way, shape, or form does the `..` impact parsing of anything.

What I mean by precedence is a more broad decision about what order things will be done given a block of code.  You literally can't interpret that code without having the precedence of deciding what order things will be done in.  So I don't know, maybe I meant to say "execution order", but I don't see why it should be separated from "precedence" because it's all about needing to know the order of operations.  Another distinction without a difference.

**Comment by CyrusNajmabadi** [comment]
>  If you're honestly suggesting I do that everywhere I use the spread operator, I've gotta assume you're being sarcastic. 

That is waht i'm suggesting.  That's our suggestion for any of the hundreds of cases where we do not support customizing formatting and a user wants to do something custom.

(Our first suggestion is simply: do what the formatter does.  But if you don't want that, then disabling is our recommendation).

> What I mean by precedence is a more broad decision about what order things will be done given a block of code.

That's not really what these terms mean for the actual language though.  And, as i pointed out above:

1. it's def misleading/confusing to think of things those ways.
2. even if you thought about things those ways, it would motivate formatting in the way we do.

> Another distinction without a difference.

Fundamentally, we see these are extremely different and relevant.  With your perspective, literally everything is an operator.  Comma becomes an operator (because you need to know the thing before the comma executes before the thing after the comma), etc. 

The language (and tools) are built out of real rules about how things are actually designed and composed.  It's very relevant, for example, that we do not think of expressions as statements, or signatures as types, etc. etc.  The distinctions are very relevant, even if some want to conflate them.  Thanks! :)

**Comment by jez9999** [comment]
Why isn't it reasonable to think of everything as an operator with precedence, though?  Yeah, the comma could be considered that way.  That's kind of how my brain thinks of code because it needs to know the order thing will execute:

```
var foo = bar.Baz(.. xyz(1 + 2 *  3), 456);
```

You need to know that the above is going to end up being:

```
(var foo) = (bar.Baz(.. (xyz((1 + (2 *  3))), 456)));
```

You could even make a case for that being the language syntax, though obviously it looks pretty unreadable (though Lisp kind of went that way...)

**Comment by CyrusNajmabadi** [comment]
> Why isn't it reasonable to think of everything as an operator with precedence, though?

Because that's not how our language is designed.  And it's not a goal to make all our tools work with any arbitrary interpretation of any particular user.  You're welcome to think things work a particular way, but we're not going to necessarily support that (just like we would not support someone who think it works in different way diametrically opposed to your own interpretation).

> You could even make a case for that being the language syntax

Sure.  That would not be our language though.  So...

> Interestingly, `,` seems to be considered an operator [in C](https://learn.microsoft.com/en-us/cpp/c-language/precedence-and-order-of-evaluation?view=msvc-170#precedence-and-associativity-of-c-operators).

Yes.  But not ours.  Intentionally.  Different languages are *different*.  That's why they exist.  They have different designs, goals and approaches.  You can attempt to use what you know from one to help intuit to anotehr.  But that intuition may be vastly incorrect.  For example, if i use what i know from C# and apply it to many other languages, i'll find myself incorrect in many cases.   Intution and collapsing ideas can be ok to try to get by.  But, fundamentally, if you want to work *well* with a particular language, you'll have to understand that language itself deeply and how it actually works.

**Comment by CyrusNajmabadi** [comment]
@jez9999 At this poitn, if you have feedback on language design, you should take it to dotnet/csharplang.  I get that you think of hte language in a particular fashion, but there's not much we can do about that here.  And it's certainly not going to make much difference here in the formatting engine extension API discussion.

**Comment by Eagle3386** [comment]
7 (in words: S-E-V-E-N!) years later & given the most disrespectful closing-comment I _ever_ read in 20 years as a professional software engineer over in issue #75607, when do we _actually_ get this API in our hands?
I _really_ want to know, just so that we can do the job of others (read: this repo's maintainers which are paid by Microsoft to improve its C# compiler) in less time, probably within less than a month.

**Comment by CyrusNajmabadi** [comment]
@Eagle3386 I'm currently working on a design document for this.

**Comment by Eagle3386** [comment]
Sure, it's been just 7 years now. But again: you're free to chose. So do I. Bye.

**Comment by CyrusNajmabadi** [comment]
@Eagle3386 please abide by the code of conduct.  Thanks.

**Comment by hodorstorm** [comment]
What is the ETA on this issue being resolved?  It is 8 years old and dozens of other issues were closed as "dups" with the promise that this issue will be quickly implmented and subsuming all those other issues.

**Comment by hodorstorm** [comment]
At least someone finally unlocked this conversation because of [this](https://github.com/dotnet/roslyn/issues/81921).

[dotnet](https://github.com/dotnet)
unlocked this conversation [4 hours ago](https://github.com/dotnet/roslyn/issues/31691#event-21926214664)

**Comment by CyrusNajmabadi** [comment]
> What is the ETA on this issue being resolved?

Unclear.  I started working on the new design a couple of months ago.  It's def large.  

> the promise that this issue will be quickly implmented 

We have not made that promise.  This (and most formatting work) is currently on the backlog.  I'm working on a design here because i really would like us to be able to make forward progress on this.  However, it is an extremely challenging space.  

Specifically, a while back, the code for roslyn was extracted into a shared lib that was then made part of dotnet-format.  It also became the case that errors were added to the dotnet tooling if one's code fell out of line with the current roslyn formatter.

That introduced a major compatibility problem for our team.  Specifically, any changes in behavior in the formatting engine are now *breaking changes* for customers (as many have turned on these rules as errors).  This is severely problematic, especially because you may have many people on a team working with varying versions of .net or VS (which can all ship different versions of the formatting engine).

This is compounded by the problem that the existing formatting engine (written ages ago by someone who left the team long ago) is extremely opaque and baroque, and has proven very difficult to modify *safely*.  Virtually all changes that people have made to it have ended breaking some set of customers somewhere.  

This has put us in the painful position (oft hated in programming) where *all* observable behavior we have is something that users have taken a dependency on, and which is considered breaking to change.  This has ossified this component and locked up being able to do effective work here.

--

I would like us to be out of this situation.   But that means designing something that is actually health/sustainable.  This means instead of massive files containing all the logic for c# interwoven in humongous methods (which are brittle and nearly impossible to tweak), we have a clearly designed and exhaustively tested engine that is flexible enough for the needs we have.  Then, the rules we want (like "i want a space between the open paren and arg in an argument list") are things that can be specified simply, declaratively, and in a way that will not interfere with other rules (unless they are for those same nodes/tokens as well).   

As you might imagine, such a design is non-trivial and takes a lot of time to writeup, validate, and work through the enormous backlog of feature requests to ensure it will be sufficient.

On top of that, since this design is happening, it's also a time to look into addressing major limitations in the existing engine.  For example, being able to have tabular formatting, or preserving tabs properly, or dealing with pp directives well, or indenting lambda/etc. well, etc. etc.  We have broad areas that people have complained about for years, and we want to solve those.  

--

>  and dozens of other issues were closed as "dups" 

Unfortunately, like many teams, we lack good resourcing for these types of issues.  We have a backlog of thousands of issues, and far too few people to work through them all.  By collecting these all under one umbrella, it greatly *increases* the chances for funding here as we can legitimately point to all the places people have filed issues/requests and show how relevant this is for the ecosystem. 

'duping' means that there is good visibility on just how important this issue is.  Having hundreds of disparate issues floating around on their own is the best way to ensure nothing happens with those, as their individual priority/impact will be too low to invest in.

**Comment by hodorstorm** [comment]
>> What is the ETA on this issue being resolved?

> Unclear. I started working on the new design a couple of months ago. It's def large.

This is exactly why it makes sense to solve the other, far simplier issues rather than waiting for this to be done.  This issue has been opened for eight years.  It's only reasonable to expect it to take several more years, and that's being optimistic.  Some of the other issues that were closed in favor of this "do everything" API would take an hour or so to fix.

> However, it is an extremely challenging space.

Exactly.  That's why [Visual Studio 2022 incorrectly formats code with parameterized types](https://github.com/dotnet/roslyn/issues/81460#top) is a far simplier and quicker fix to make.  Hence it should not be closed as a duplicate.  That issue askes for something small and easy.  This issue asks for something very large and difficult that does a lot more things and doesn't specifically address what 81460 asks for.  Why delay an easy fix for something grandeous?
 
> Specifically, any changes in behavior in the formatting engine are now breaking changes for customers (as many have turned on these rules as errors). 

Make the default rules backwards compatible, but let the end user opt-in for other rules.

> Unfortunately, like many teams, we lack good resourcing for these types of issues. 

A good reason to do the small fixes rather than trying to architect and entire API for formatting.

> 'duping' means that there is good visibility on just how important this issue is. Having hundreds of disparate issues floating around on their own is the best way to ensure nothing happens with those, as their individual priority/impact will be too low to invest in.

A duplicate is an issue that describes the same problem or solution as another issue.  It's not an issue that could be sidestepped by implementing a far larger change than the fix proposed in the issue itself.  The problem with generalizing too much is that you spend more time on the general solution -- which also is more complicated to use -- than you would fixing all the little issues.  And all of those issues you marked as "dups" become issues stated as "using the API you created, how do I do [thing that I described in the "duped" issue]?"

Ultimately, people don't want a formatting API.  They want a simple dialog for setting rules like in JetBrains ReSharper's formatter, and they want to extract those rules into a .editorconfig file to enforce coding standards.

**Comment by CyrusNajmabadi** [comment]
> This is exactly why it makes sense to solve the other, far simplier issues

I explained why that's not possible. 

> Make the default rules backwards compatible, but let the end user opt-in for other rules.

The existing formatting engine doesn't work that way.  That's what I was saying. It's why we need new a new system that would allow for that. 

> A good reason to do the small fixes

The fixes are not small. As I've explained (across numerous issues at this point), we tried exactly that. And it was extremely problematic, leading to painful rollbacks.  This made things even more painful as you can have disparate customer segments on the different releases, which now find their tooling breaking depending on machine/user. 

> They want a simple dialog

Trying to do adhoc fixes, while breaking lots of users, will not get you that. 

There is nothing about a properly designed API that precludes a good UI. Indeed, that's part of the design. 

But we have to be in a state where we can ship safely without breakage.  That's why nothing is simple right now, and why attempts to do that for *several years* after dotnet format shipped all had to be rolled back. 

> would take an hour or so to fix.

This is not the case. And repeating it is not helping the discussion. We have repeatedly tried going this route and we've had to repeatedly roll back the changes because they invariably break some unintended case for some user. 

It is also not sufficient to simply make things "opt in". Users will opt in, and then find something else broken. Or we will find we need to correct something in this rule due to some bad interaction with something else later. Which will then break the users that opted in.

**Comment by miloush** [comment]
> people don't want a formatting API

I want a formatting API

**Comment by hodorstorm** [comment]
>> people don't want a formatting API

> I want a formatting API

You are the exception, not the rule.  Most people just want to be able to easily configure the rules and have an adequate base set of rules to cover every common desire.

I'm not opposing the creation of a formatting API.  I am opposing holding off the creation of simple, common, and obvious rules for years in order to develop that formatting API.  Even if the formatting API exists, most people will want and expect to see spacing rules for angle brackets along side the spacing rules for square and round brackets in the formatting options dialog.

If I don't have to write a rule through an API to get a space before a square bracket or a round bracket, why should I have to do that for an angle bracket?  That's inconsistent.

The formatting API should _augment_ common rules, not replace them.  Basic formatting should not require writing code against the API.  Only advanced and unforeseeable rules should require this.  Keep it simple.

**Comment by CyrusNajmabadi** [comment]
> why should I have to do that

You don't have to do that. That's the point. The point is that this system would allow anyone to do this (including us) safely. 

Right now we lack that.  Because all formatting logic is in a massive l, intertwined, baroque pile of logic that is brittle and constantly breaks unintended scenarios when touched. 

The idea behind the next formatting engine *based upon an actual API designed for the purpose* is so that is no longer a problem.

**Comment by CyrusNajmabadi** [comment]
> Basic formatting should not require writing code against the API.

I strongly disagree.  We need to ensure the API solves the core cases. Otherwise we're right back here, we're adding new code always breaks things.

**Comment by rjgotten** [comment]
> @hodorstorm 
> Ultimately, people don't want a formatting API. They want a simple dialog for setting rules like in JetBrains ReSharper's formatter, and they want to extract those rules into a .editorconfig file to enforce coding standards.
> [..]
> Most people just want to be able to easily configure the rules and have an adequate base set of rules to cover every common desire.

And guess what the quickest way way to get to a situation having a broad ruleset capable of covering every common desire, actually _is_? That's right- having a publicly accessible formatting API so you can community-source the formatting rules. And ensure the community can keep them up to date, patching them for e.g. newly added language features that introduce new syntax not quite compatible with the existing rules of _some_ of those common desires, with some additional love and attention.

Which Microsoft - very clearly illustrated by how long certain of these issues linger on the backlog - **doesn't have**, _even_ if they'd want to.

**Comment by hodorstorm** [comment]
> And guess what the quickest way way to get to a situation having a broad ruleset capable of covering every common desire, actually is? That's right- having a publicly accessible formatting API so you can community-source the formatting rules.

If it's so quick, then why is rhis issue opened doe EIGHT YEARS and development has not even started.  Your statement is empirically false.  There is no way it would have taken eight years to put an option in the existing formatter to add a space before opening angle brackets.  If it takes more than five minutes to add that option, you're doing something wrong.

But hey, what's the ETA on this API?  Will it be done this year?

**Comment by CyrusNajmabadi** [comment]
> If it's so quick

It's quicker and safer than the alternatives.

> EIGHT YEARS and development has not even started. 

We have issues opened longer.  Time to get to something depends on priority, resourcing, and other team objectives. 

> There is no way it would have taken eight years to put an option in the existing formatter to add a space before opening angle brackets. 

Putting a new option takes *longer* as we have demonstrated many times that there is a very high chance of regressions, which then is often extremely painful for the community and requires painful rollbacks.

> If it takes more than five minutes to add that option, you're doing something wrong.

Many people have thought that, tried to just add new rules, with new options, and run into problems immediately.  Even testing this is enormously difficult as *every permutation* of this rule with all existing rules needs to be tested and validated.

Remember, even if this is opt in, once shipped *the behavior* in any config *cannot change*.  Otherwise you can and will break real codebases.  This effectively also doubles the entire testing surface area (which is already underserved by existing tests).  Every single test now needs a variant that checks it with this option on, and validates the behavior of the rules together, and ensures no regressions there in the future.

This is much more than 5 minutes.  It's easily thousands and thousands of tests.  Otherwise regressions can and will happen.  Again, we have tried this numerous times.  And there are dozens of cases where this broke real world deployments.  

> But hey, what's the ETA on this API? Will it be done this year?

Can't say.  It will depend on resourcing and prioritization.  In the meantime, if this is critical to you, you can always just write your own tree transformer to do this and use that yourself.  That's why roslyn exposes an api for transformation.  So you are not blocked on us and our limited resources, to scratch an itch you have.  If this is critical for you, you can always invest in it yourself.

**Comment by hodorstorm** [comment]
https://github.com/dotnet/roslyn/issues/81460
Visual Studio 2022 incorrectly formats code with parameterized types

I fixed this issue since the people at Microsoft can't.  Just run the following extension installation.  It took literally one line of code.

I can't post the fix on the original post because Microsoft shut down the conversation there.

Anyway, enjoy.  Now we don't have to wait another ten years for the bracket formatting fix.

[FormatProperly.zip](https://github.com/user-attachments/files/26319966/FormatProperly.zip)

**Comment by hodorstorm** [comment]
I love how all of CyrusNajmabadi's buddies come to just downvote the solution to this problem.  This high-school click behavior is exactly why online communties fail and become echo-chambers.

**Comment by CyrusNajmabadi** [comment]
@hodorstorm hunh?  I have no idea who those people are. Why do you think they are buddies of mine?

**Comment by hodorstorm** [comment]
Yeah, right.

**Comment by XeonG** [comment]
> [#81460](https://github.com/dotnet/roslyn/issues/81460) Visual Studio 2022 incorrectly formats code with parameterized types
> 
> I fixed this issue since the people at Microsoft can't. Just run the following extension installation. It took literally one line of code.
> 
> I can't post the fix on the original post because Microsoft shut down the conversation there.
> 
> Anyway, enjoy. Now we don't have to wait another ten years for the bracket formatting fix.
> 
> [FormatProperly.zip](https://github.com/user-attachments/files/26319966/FormatProperly.zip)

Why not put the source somewhere or add it to the package system, just a bit weird dumping .zip with .vsix extension in todays scam riddled world.

Though I am perplexed why this issue has been left to stagnate under ooh we need a proper formatting or we break some codebases really? A specific csharp rule for not indenting #region/endregion in line the the rest of code is something that would break millions of codebases and need a gazallion test cases.


> Remember, even if this is opt in, once shipped the behavior in any config cannot change. Otherwise you can and will break real codebases. This effectively also doubles the entire testing surface area (which is already underserved by existing tests). Every single test now needs a variant that checks it with this option on, and validates the behavior of the rules together, and ensures no regressions there in the future.


What a load, I literally couldn't give a shit, MS employees making out that they care about not screwing up things,.. meanwhile Winshit11 is a pure case of screwing up on so many levels as is so many other things they do these days.


> duplicate is an issue that describes the same problem or solution as another issue. It's not an issue that could be sidestepped by implementing a far larger change than the fix proposed in the issue itself. The problem with generalizing too much is that you spend more time on the general solution -- which also is more complicated to use -- than you would fixing all the little issues. And all of those issues you marked as "dups" become issues stated as "using the API you created, how do I do [thing that I described in the "duped" issue]?"

> Ultimately, people don't want a formatting API. They want a simple dialog for setting rules like in JetBrains ReSharper's formatter, and they want to extract those rules into a .editorconfig file to enforce coding standards.

yeah basically

**Comment by CyrusNajmabadi** [comment]
> I literally couldn't give a ****

Then we are at an impasse.  We care about breaking changes.  It is extremely painful, and very costly for companies and organizations pulling in net .net versions.  We do not take them lightly and every time we make changes here it does take an enormous amount of testing to validate things.  as mentioned in the thread @XeonG, a pp directive is one of hte harder changes to make.  IT impacts *every possible other formatting rule*.  Because a pp directive can appear between all pairs of tokens in every possible node case.  so you both need to validate it and validate that it works with *all* other rules.  But you also then can't change that behavior from then on.  If you do, that is a breaking change and will almost certainly impact someone.

**Comment by sharwell** [comment]
Source code formatting is the hardest problem I've ever worked on in compilers/languages, where 'hard' is defined by (actual difficulty)/(expected difficulty). The technical difficulty is only half the problem. The other half the problem comes from the fact that there is simply no way to understand just how difficult the problem is until you actually try and create one to the standards Visual Studio aims for. I've had people who spent their whole career working on compiler _front-ends_ (probably the closest career path not directly responsible for formatting) tell me it's crazy to take that long to get a formatter working properly. Personally, I was about a year in working on my first full formatter before I really came to accept this (many years before I ever worked on the Roslyn team).

When I was on Roslyn, I never fully committed to the extensible formatter because I never saw a path to completing it inside an annual review cycle, and it always seemed to be way too high a risk that the results wouldn't be visible in a short enough time that the project stayed funded. It's too difficult a task to explain why it's difficult to people deciding what your career will be like in following years.

**Comment by Eagle3386** [comment]
@sharwell:
> It's too difficult a task to explain why it's difficult to people deciding what your career will be like in following years.

Well, either exchange those people or make everybody leave those people on their own - some have to learn it the hard way they're toxic & then either change themselves or ultimately die alone.
Such awful people shall never be in charge. Ever. And yes, we - and only we - have the power to change that. We are many more than those puny couple of people!

**Comment by XeonG** [comment]
I dunno guys it just seems like everything that is formatting related has been tied up into this extensible formatter which is too big a thing.

And what is a pp directive?

@sharwell Well I get what you're saying but that is unfortunate it has been setup that way

**Comment by miloush** [comment]
> And what is a pp directive?

[Preprocessor directives](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives)

**Comment by CyrusNajmabadi** [comment]
> I dunno guys it just seems like everything that is formatting related has been tied up into this extensible formatter which is too big a thing.

There are no small things here unfortunately.  

> And what is a pp directive?

A preprocessor directive.  These are particularly hard because literally all of them can appear between any two tokens in literally any node construct whatsoever.  the testing/validation matrix for them is enormous.  And it is trivial to accidentally break it, which will be breaking change for the consumer.

**Comment by CyrusNajmabadi** [comment]
> Well I get what you're saying but that is unfortunate it has been setup that way

Fundementally, this space is *enormously* expensive, and has a massive set of users who are highly intolerant to *any* change in behavior.  This makes it a super high risk.  I've been in this space a long time, and I have a hard time convincing myself that it's even worth investing that much.

**Comment by hodorstorm** [comment]
[A timely video](https://www.youtube.com/watch?v=t992ul_IKtc) that is very pertinent to this discussion.  Ah, if only we applied these lessons today.

**Comment by CyrusNajmabadi** [comment]
I great thing about Roslyn is that it is open source. So if there's something you want it to do, and you have your own idea of what is ok or not in your domain, you have the total power to go make it do that for yourself, and whoever you want to give it to :-)

**Comment by vdogamr** [comment]
How about this as a suggestion for all the indentation subtasks here.  Just leave them where I put them and don't move them.  Give me an option to turn off any re-alignment of my preproccessor directives and comments.
Note, I am only here because I care about `#region` and `#endregion`.  Maybe the rest of these things are difficult to implement, but these two seem like they should be trivial.  Especially if the answer is to just not move them at all.
